JIT: Add scalar evolution analysis and do IV widening based on it

jakobbotsch opened the pull request:
This adds a new phase meant for optimizing induction variables. It adds infrastructure for SSA-based analysis of induction variables (scalar evolution analysis), and uses it to do induction variable widening.
For example, with this optimization, codegen for
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static int Foo(int[] arr)
{
    int sum = 0;
    for (int i = 0; i < arr.Length; i++)
    {
        sum += arr[i];
    }

    return sum;
}
```
goes from
```asm
; Assembly listing for method ConsoleApp34.Program:Foo(int[]):int (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; FullOpts code
; optimized code
; rsp based frame
; fully interruptible
; No PGO data
; Final local variable assignments
;
;  V00 arg0         [V00,T02] (  4,  7   )     ref  ->  rcx         class-hnd single-def <int[]>
;  V01 loc0         [V01,T01] (  4, 10   )     int  ->  rax        
;  V02 loc1         [V02,T00] (  5, 17   )     int  ->  rdx        
;  V03 OutArgs      [V03    ] (  1,  1   )  struct (32) [rsp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;  V04 cse0         [V04,T03] (  3,  6   )     int  ->   r8         "CSE - aggressive"
;
; Lcl frame size = 40

G_M8112_IG01:
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25
G_M8112_IG02:
       xor      eax, eax
       xor      edx, edx
       mov      r8d, dword ptr [rcx+0x08]
       test     r8d, r8d
       jle      SHORT G_M8112_IG04
       align    [0 bytes for IG03]
						;; size=13 bbWeight=1 PerfScore 3.75
G_M8112_IG03:
       mov      r10d, edx
       add      eax, dword ptr [rcx+4*r10+0x10]
       inc      edx
       cmp      r8d, edx
       jg       SHORT G_M8112_IG03
						;; size=15 bbWeight=4 PerfScore 19.00
G_M8112_IG04:
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25

; Total bytes of code 37, prolog size 4, PerfScore 24.25, instruction count 14, allocated bytes for code 37 (MethodHash=d1cce04f) for method ConsoleApp34.Program:Foo(int[]):int (FullOpts)
; ============================================================
```
to
```asm
; Assembly listing for method ConsoleApp34.Program:Foo(int[]):int (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; FullOpts code
; optimized code
; rsp based frame
; fully interruptible
; No PGO data
; Final local variable assignments
;
;  V00 arg0         [V00,T02] (  4,  7   )     ref  ->  rcx         class-hnd single-def <int[]>
;  V01 loc0         [V01,T01] (  4, 10   )     int  ->  rax        
;* V02 loc1         [V02,T04] (  0,  0   )     int  ->  zero-ref   
;  V03 OutArgs      [V03    ] (  1,  1   )  struct (32) [rsp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;  V04 tmp1         [V04,T00] (  5, 17   )    long  ->   r8         "Widened primary induction variable"
;  V05 cse0         [V05,T03] (  3,  6   )     int  ->  rdx         "CSE - aggressive"
;
; Lcl frame size = 40

G_M8112_IG01:  ;; offset=0x0000
       sub      rsp, 40
						;; size=4 bbWeight=1 PerfScore 0.25
G_M8112_IG02:  ;; offset=0x0004
       xor      eax, eax
       mov      edx, dword ptr [rcx+0x08]
       test     edx, edx
       jle      SHORT G_M8112_IG04
       xor      r8d, r8d
       align    [0 bytes for IG03]
						;; size=12 bbWeight=1 PerfScore 3.75
G_M8112_IG03:  ;; offset=0x0010
       add      eax, dword ptr [rcx+4*r8+0x10]
       inc      r8d
       cmp      edx, r8d
       jg       SHORT G_M8112_IG03
						;; size=13 bbWeight=4 PerfScore 18.00
G_M8112_IG04:  ;; offset=0x001D
       add      rsp, 40
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25

; Total bytes of code 34, prolog size 4, PerfScore 23.25, instruction count 13, allocated bytes for code 34 (MethodHash=d1cce04f) for method ConsoleApp34.Program:Foo(int[]):int (FullOpts)
```
where we were able to drop a zero extension of the index inside the loop. In the future I plan to build strength reduction on top of the same analysis package.

The analysis is inspired by [1] and by LLVM's scalar evolution package. It provides a small IR that represents the evolving value of IR nodes inside loops. At the core of this IR is the notion of an "add recurrence", which describes a changing value as `<loop, start, step>`; the value of such an add recurrence is $start + N * step$, where N is the iteration index. Currently only simple add recurrences are supported where the start and step are either constants or invariant locals, but the framework generalizes nicely to allow chains of recurrences if we wish to support that. The IR also supports constants, invariant locals and operators on top of these (casts, adds, multiplications and shifts).

<details>
<summary>Analysis example</summary>

For the IR for the above, the analysis produces the following:
```scala
Analyzing scalar evolution in L00 header: BB03
  Members (1): BB03
  Entry: BB02 -> BB03
  Exit: BB03 -> BB04
  Back: BB03 -> BB03
BB03 [0001] [006..016) -> BB03,BB04 (cond), preds={BB02,BB03} succs={BB04,BB03}

STMT00009 ( ??? ... ??? )
N004 (  0,  0) [000045] DA---------                         ▌  STORE_LCL_VAR int    V01 loc0         d:3
N003 (  0,  0) [000044] -----------                         └──▌  PHI       int   
N001 (  0,  0) [000050] ----------- pred BB03                  ├──▌  PHI_ARG   int    V01 loc0         u:4
N002 (  0,  0) [000047] ----------- pred BB02                  └──▌  PHI_ARG   int    V01 loc0         u:2

[000047] => 0

STMT00007 ( ??? ... ??? )
N004 (  0,  0) [000041] DA---------                         ▌  STORE_LCL_VAR int    V02 loc1         d:3
N003 (  0,  0) [000040] -----------                         └──▌  PHI       int   
N001 (  0,  0) [000051] ----------- pred BB03                  ├──▌  PHI_ARG   int    V02 loc1         u:4
N002 (  0,  0) [000048] ----------- pred BB02                  └──▌  PHI_ARG   int    V02 loc1         u:2

[000051] => <L00, 1, 1>
[000048] => 0
[000041] => <L00, 0, 1>
```
For example, here it was able to show that `V02`, the index, is a primary induction variable; it is an add recurrence that starts at 0 and steps by 1 every iteration of the loop. It also showed that the value that comes from the backedge is similarly an add recurrence, except that it starts at 1 in the first loop.

```scala
STMT00003 ( 0x006[E-] ... 0x00B )
N015 (  8,  9) [000015] DA--GO-----                         ▌  STORE_LCL_VAR int    V01 loc0         d:4
N014 (  8,  9) [000014] ----GO-----                         └──▌  ADD       int   
N012 (  6,  7) [000033] ----GO-N---                            ├──▌  COMMA     int   
N001 (  0,  0) [000025] -----------                            │  ├──▌  NOP       void  
N011 (  6,  7) [000034] n---GO-----                            │  └──▌  IND       int   
N010 (  3,  5) [000032] -----O-----                            │     └──▌  ARR_ADDR  byref int[]
N009 (  3,  5) [000031] -------N---                            │        └──▌  ADD       byref 
N002 (  1,  1) [000022] -----------                            │           ├──▌  LCL_VAR   ref    V00 arg0         u:1
N008 (  4,  5) [000030] -------N---                            │           └──▌  ADD       long  
N006 (  3,  4) [000028] -------N---                            │              ├──▌  LSH       long  
N004 (  2,  3) [000026] ---------U-                            │              │  ├──▌  CAST      long <- uint
N003 (  1,  1) [000023] -----------                            │              │  │  └──▌  LCL_VAR   int    V02 loc1         u:3
N005 (  1,  1) [000027] -------N---                            │              │  └──▌  CNS_INT   long   2
N007 (  1,  1) [000029] -----------                            │              └──▌  CNS_INT   long   16
N013 (  1,  1) [000009] -----------                            └──▌  LCL_VAR   int    V01 loc0         u:3 (last use)

[000022] => V00.1
[000023] => <L00, 0, 1>
[000026] => <L00, 0, 1>
[000027] => 2
[000028] => <L00, 0, 4>
[000029] => 16
[000030] => <L00, 16, 4>
[000031] => <L00, (V00.1 + 16), 4>
[000032] => <L00, (V00.1 + 16), 4>
```
This one is more interesting since we can see hints of how strength reduction is going to utilize the information. In particular, the analysis was able to show that the address `[000032]` is also an add recurrence; it starts at value `(V00.1 + 16)` (the address of the first array element) and steps by `4` in every iteration.

```scala
STMT00004 ( 0x00C[E-] ... 0x00F )
N004 (  3,  3) [000019] DA---------                         ▌  STORE_LCL_VAR int    V02 loc1         d:4
N003 (  3,  3) [000018] -----------                         └──▌  ADD       int   
N001 (  1,  1) [000016] -----------                            ├──▌  LCL_VAR   int    V02 loc1         u:3 (last use)
N002 (  1,  1) [000017] -----------                            └──▌  CNS_INT   int    1

[000016] => <L00, 0, 1>
[000017] => 1
[000018] => <L00, 1, 1>
[000019] => <L00, 1, 1>

STMT00002 ( 0x010[E-] ... 0x014 )
N005 (  7,  7) [000008] ---X-------                         ▌  JTRUE     void  
N004 (  5,  5) [000007] J--X---N---                         └──▌  GT        int   
N002 (  3,  3) [000006] ---X-------                            ├──▌  ARR_LENGTH int   
N001 (  1,  1) [000005] -----------                            │  └──▌  LCL_VAR   ref    V00 arg0         u:1
N003 (  1,  1) [000004] -----------                            └──▌  LCL_VAR   int    V02 loc1         u:4

[000005] => V00.1
[000004] => <L00, 1, 1>
```

Here the analysis shows that the array object is invariant (otherwise analysis would fail) and that the compared index is an add recurrence that starts at 1.

</details>

When the induction variable has uses after the loop the widening pass will store the widened version back to the old local. This is only possible if all exits where the old local is live-in are not critical blocks in the sense that all their preds must come from inside the loop.
Exit canonicalization ensure sthat this is usually the case, but RBO/assertion prop may have uncovered new natural loops, so we still have to repeat the check.

[1] Michael Wolfe. 1992. Beyond induction variables. In Proceedings of the ACM SIGPLAN 1992 conference on Programming language design and implementation (PLDI '92). Association for Computing Machinery, New York, NY, USA, 162–174. https://doi.org/10.1145/143095.143131

jakobbotsch committed:
WIP

jakobbotsch committed:
WIP

jakobbotsch committed:
Check invariance on IR instead of on SCEV to avoid cycles

jakobbotsch committed:
Some work on recursive scevs

jakobbotsch committed:
Simplify for now

jakobbotsch committed:
IV widening, fix bugs

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into simplify-ivs

jakobbotsch committed:
Bail for non-integer constants

jakobbotsch committed:
Add costing

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Clean up

jakobbotsch committed:
Fix linux build

jakobbotsch committed:
Clean up

jakobbotsch commented:
/azp run runtime-coreclr jitstress, runtime-coreclr libraries-jitstress, Fuzzlyn

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into simplify-ivs

jakobbotsch commented:
Test failures look like #97892.
Regressions are much larger outside win-x64. Need to understand why, when I look at the examples almost all of them are range check related for the reason I mentioned above. I don't understand why that wouldn't affect win-x64 equally, though.
Some of the arm64 diffs look to be because we do not support the zero-cost `STORE_LCL_VAR(CAST<long <- int>(data))` pattern there, but that should be implementable. Also, on arm64 the IV widening is proportionally less useful since address modes can zero extend as part of the instruction, so fewer direct benefits from IV widening are to be expected.

jakobbotsch committed:
JIT: Support more scaled addressing modes on arm64

We currently support scaled addressing modes when the index also needs
an extension through contained `BFIZ` nodes. However, we did not support
scaled addressing modes if the index was 64 bits. This adds that
support as a natural extension to the `GT_LEA`.

jakobbotsch committed:
Do not swap addrecs unnecessarily when folding

jakobbotsch committed:
Add some todos

jakobbotsch committed:
Move phase later; compute loops; address some TODOs

jakobbotsch committed:
Add docs, enhance optIsIVWideningProfitable when zero extensions are sources of stores

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Remove unnecessary config var

jakobbotsch commented:
There's a number of cases where it would also be beneficial to widen uses of the narrowed IV variable after the loop.
In theory it should be possible with some additional dominance check, but it also requires us to go find those uses which in general I'm a bit wary off due to TP.

One example from our benchmarks:
https://github.com/dotnet/performance/blob/d8f4249eb1c94aaa255df87c894082363be96856/src/benchmarks/micro/runtime/Benchstones/BenchI/QuickSort.cs#L34-L44
It would be beneficial to widen in the three array subscripts after the last inner loop.

jakobbotsch committed:
Merge branch 'main' into simplify-ivs

AndyAyersMS commented:
> When the induction variable has uses after the loop the widening pass will store the widened version back to the old local. This is only possible if all exits where the old local is live-in are not critical blocks in the sense that all their preds must come from inside the loop.
> This practically disqualifies all such loops because we do not canonicalize exits today, and usually the regular exit blocks have a non-loop predecessor because of loop inversion. In the future we can solve this problem by canonicalizing exit blocks such that they only have loop preds.

Another less elegant strategy here would be to just have two IVs, the original and the widened one.

jakobbotsch committed:
JIT: Canonicalize loop exits

This adds another canonicalization requirement for loops during the
optimization phases, namely that all regular loop exit blocks have only
loop predecessors. This gives a natural place to insert IR that we want
to run only when we know the loop was entered.

Exceptional loop exit blocks can still have non-loop predecessors, so
these must still be accounted for by optimizations.

jakobbotsch committed:
Remove some manual canonicalization code

jakobbotsch committed:
Remove unused declaration

jakobbotsch committed:
Clean up

jakobbotsch committed:
Fix function header

jakobbotsch committed:
Handle x86; skip regular exit blocks that are handlers

jakobbotsch committed:
Set proper weight of exit

jakobbotsch committed:
Add stats

jakobbotsch committed:
Merge remote-tracking branch 'origin/canonicalize-loop-exits' into simplify-ivs

jakobbotsch commented:
> Another less elegant strategy here would be to just have two IVs, the original and the widened one.

That would certainly also be possible, though not sure whether it's still worth it to do widening then.

Some stats of number of widenings we do over win-x64 collections:
asp.net:
```scala
Num widenings
     <=          0 ===>   64149 count ( 94% of total)
      1 ..       1 ===>    2088 count ( 98% of total)
      2 ..       2 ===>     956 count ( 99% of total)
      3 ..       3 ===>     221 count ( 99% of total)
      4 ..       4 ===>      47 count ( 99% of total)
      5 ..       5 ===>      14 count ( 99% of total)
      6 ..       6 ===>      23 count ( 99% of total)
      7 ..       7 ===>      29 count ( 99% of total)
      8 ..       8 ===>       2 count ( 99% of total)
      9 ..       9 ===>       1 count ( 99% of total)
     10 ..      10 ===>       2 count (100% of total)
      >         10 ===>       7 count (100% of total)
```

benchmarks.run_pgo:
```scala
Num widenings
     <=          0 ===>   35832 count ( 89% of total)
      1 ..       1 ===>    3008 count ( 97% of total)
      2 ..       2 ===>     558 count ( 98% of total)
      3 ..       3 ===>     220 count ( 99% of total)
      4 ..       4 ===>     178 count ( 99% of total)
      5 ..       5 ===>       9 count ( 99% of total)
      6 ..       6 ===>       5 count ( 99% of total)
      7 ..       7 ===>       1 count ( 99% of total)
      8 ..       8 ===>       7 count ( 99% of total)
      9 ..       9 ===>       6 count ( 99% of total)
     10 ..      10 ===>       4 count (100% of total)
      >         10 ===>      24 count (100% of total)
```

With #98096 also included:
asp.net
```scala
Num widenings
     <=          0 ===>   63847 count ( 94% of total)
      1 ..       1 ===>    2269 count ( 97% of total)
      2 ..       2 ===>    1020 count ( 99% of total)
      3 ..       3 ===>     226 count ( 99% of total)
      4 ..       4 ===>      84 count ( 99% of total)
      5 ..       5 ===>      14 count ( 99% of total)
      6 ..       6 ===>      21 count ( 99% of total)
      7 ..       7 ===>      30 count ( 99% of total)
      8 ..       8 ===>      17 count ( 99% of total)
      9 ..       9 ===>       1 count ( 99% of total)
     10 ..      10 ===>       1 count (100% of total)
      >         10 ===>       9 count (100% of total)
```

benchmarks.run_pgo:
```scala
Num widenings
     <=          0 ===>   35628 count ( 89% of total)
      1 ..       1 ===>    2966 count ( 96% of total)
      2 ..       2 ===>     756 count ( 98% of total)
      3 ..       3 ===>     267 count ( 99% of total)
      4 ..       4 ===>     177 count ( 99% of total)
      5 ..       5 ===>      11 count ( 99% of total)
      6 ..       6 ===>       5 count ( 99% of total)
      7 ..       7 ===>       1 count ( 99% of total)
      8 ..       8 ===>       7 count ( 99% of total)
      9 ..       9 ===>       6 count ( 99% of total)
     10 ..      10 ===>       4 count (100% of total)
      >         10 ===>      24 count (100% of total)
```

(so around ~10% more widening done when we include exit canonicalization)

jakobbotsch committed:
Remove stats

jakobbotsch commented:
If I change the heuristic to account for the fact that arm64 can fold zero/sign extension into the addressing modes, and also hack in a fix for the issue around unnecessary zero/sign extensions on 32-bit mutations of 64-bit variables, then the win-arm64 diffs look like this: https://gist.github.com/jakobbotsch/6c13cc63d967fb5f8cfce8946f4f902e

Overall it doesn't seem worth it to enable IV widening for arm64. Perhaps if we were a bit smarter around widening uses after the loop also and around widening in the face of sign extensions this would be different. We can reevaluate if/when we end up with these improvements in the future.

jakobbotsch committed:
Only do IV widening on x64

jakobbotsch committed:
Clean up

jakobbotsch committed:
Run jit-format

AndyAyersMS commented:
Do you have a sense of how many opportunities we still might be missing out on? Seems like the number should be fairly small, but it might be hard to get an accurate count.

jakobbotsch commented:
> Do you have a sense of how many opportunities we still might be missing out on? Seems like the number should be fairly small, but it might be hard to get an accurate count.

I think there's three main sources:
- Unnatural loops made natural by RBO/assertion prop, or made uncanonical (would be possible to handle, but gets messy since canonicalization inserts new blocks, which won't have valid liveness data without some fixups). On ASP.NET where we widen 6021 IVs there's 45 cases rejected for this reason.
- Primary IVs that the framework doesn't handle. I generalized the "primary IVs" to more general PHI cycles ([code](https://gist.github.com/jakobbotsch/6afa7604dcc6455ea7d9aafe6dbf79bd)) and we widened 7974 (as compared to the 6021) primary IVs. Not clear to me how many of these cycles are simple extensions to the SCEV analysis and how many of them are truly unrepresentable.
- Locals that are live into exceptional exits where we cannot store the wide IV back (these are DNER due to being live-in to EH and multi-def, however, so we don't consider them profitable anyway)

(2) would be interesting to look further into.

jakobbotsch commented:
Of course the above is almost exactly similar to just doing the profitability/legality checks for all phis in the header. If we do that we widen 7991 variables in asp.net, and we wouldn't need any form of scalar evolution analysis at all for this particular optimization. The [size diffs](https://gist.github.com/jakobbotsch/6315c8b34e0f8fcf97db7fc47757c99b) are worse than this PR, which is probably to be expected since those are less likely to be induction variables in the traditional sense and are much more likely to have uses after the loop.

I'll probably keep it as is, but those ~2000 cases would be good to look into and figure out if they should be modeled by the SCEV framework or not.

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into simplify-ivs

jakobbotsch committed:
Only look at regular exits

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into simplify-ivs

jakobbotsch committed:
Add depth limit; small clean ups

jakobbotsch committed:
Revert unnecessary change

jakobbotsch committed:
Remove counter

jakobbotsch committed:
Add more info for LclVarDsc reason

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into simplify-ivs

jakobbotsch commented:
Size regressions on linux-x64 are large now. Also seeing some more size regressions on win-x64 now. It's likely because of exit canonicalization allowing more cases.. One frequent case where there's a size regression (but perfscore improvement) is when we've cloned a loop; in those cases when we widen (usually in both the hot and cold loops) we create an extra place to initialize the IV variable. Previously both loops would share the initialization of the IV, while now they each get a copy of the initialization.

jakobbotsch commented:
The above can be (partly) mitigated by sharing widened IV variables that have the same reaching def for the start value (and ensuring that we then insert initialization in that block instead of in the preheader).

jakobbotsch committed:
Share widened IVs with same reaching def from outside the loop

jakobbotsch committed:
Revert regressions only change

jakobbotsch committed:
Also try replacing uses before the loop

jakobbotsch committed:
Skip cases where the same local is the IV in multiple loops

jakobbotsch committed:
More experimenting

jakobbotsch committed:
Improve

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Clean up

jakobbotsch committed:
Clean up

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Insert after phis correctly

jakobbotsch committed:
Squashed

jakobbotsch committed:
Merge remote-tracking branch 'origin/jit-metrics-squashed' into simplify-ivs

jakobbotsch committed:
Fix

jakobbotsch committed:
Print PerfScore geomeans

jakobbotsch committed:
Add to report

jakobbotsch committed:
Clean up

jakobbotsch committed:
Move overall perfscore change to details

jakobbotsch committed:
Merge branch 'main' of github.com:dotnet/runtime into simplify-ivs

jakobbotsch committed:
Fixes after merge

jakobbotsch committed:
Run jit-format

jakobbotsch committed:
Do not pick start BB for parameters

jakobbotsch commented:
linux-x64 difs are worse than windows-x64, and it shows up by having a much larger number of spilled primary IVs. The primary reason is that on linux-x64 we have much fewer callee saves available; we have just 5 (when you ignore `rbp` that is almost always used for frame chaining). On windows-x64 we have 8 callee save registers. So there's a much higher probability for primary IVs to end up spilled when the loop has a call in it. In the future we should look into what can be done in LSRA to try harder to ensure that primary IVs end up in registers (perhaps #8019 would be what to implement, or just do something more implicit the first time we see a block in a loop).

I did some automated analysis of the regressions of benchmarks.run_pgo and they are overwhelmingly from cases where we have a spilled primary IV. Furthermore, it is _very_ rare in those cases that the narrow IV was enregistered before. Based on this I don't think we need to worry about the regressions.

linux-x64 benchmarks.run_pgo:
3336 with larger size, in 3248 of these we have a spilled widened IV, in 6 the old narrow one wasn't spilled
1036 with worse perfscore, in 961 of these we have a spilled widened IV, in 5 the old narrow one wasn't spilled
9 where we spill a widened IV where the narrow IV wasn't spilled before

windows-x64 benchmarks.run_pgo:
178 with larger size, in 110 of these we have a spilled widened IV, in 0 the old narrow one wasn't spilled
115 with worse perfscore, in 65 of these we have a spilled widened IV, in 0 the old narrow one wasn't spilled
0 where we spill a widened IV where the narrow IV wasn't spilled before

jakobbotsch commented:
/azp run runtime-coreclr jitstress, runtime-coreclr libraries-jitstress

jakobbotsch commented:
cc @dotnet/jit-contrib PTAL @BruceForstall @AndyAyersMS 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=575605&view=ms.vss-build-web.run-extensions-tab). I spent a long time analyzing these, see https://github.com/dotnet/runtime/pull/97865#issuecomment-1927207885 and https://github.com/dotnet/runtime/pull/97865#issuecomment-1959294725.

There were a bunch of regressions when IV widening in cases where the same local is the primary IV of multiple loops. I've disabled this case for now; I want to look closer at the case and figure out how to enable it separately.

The initial implementation always placed the initialization of the widened IV in the preheader. However, if the reaching def from outside the loop comes from a block much further back, then that means the lifetime of the widened IV is shorter than the old IV. In a lot of cases that resulted in suboptimal register allocation of the widened IV (because other live variables ended up taking the "good" register before we got to the initialization of the widened IV). So this implementation tries to place the initialization of the widened IV in the same spot as the old narrow IV, and then does a best-effort pass to replace uses of the narrow IV with the wide IV on the way to the loop. That seems to work well, but it's a bit of a hack -- I hope to reevaluate it in the future, perhaps with some LSRA work.

Throughput impact is in the 0.2-0.3% range for most collections, with benchmarks.run_pgo being a bit worse.

AndyAyersMS commented on src/coreclr/jit/inductionvariableopts.cpp:
```diff
@@ -0,0 +1,1533 @@
+//
+// Parameters:
+//   block - Block containing the tree
+//   tree  - Tree node
```
Add comment for `depth`

AndyAyersMS commented on src/coreclr/jit/inductionvariableopts.cpp:
```diff
@@ -0,0 +1,1533 @@
+            // => <L, 0, <L, 1, 1>>
+            //
+            // The main issue is that it requires cache invalidation afterwards
+            // and turning the recursive result into an addrec.
```
Any idea how prevalent these more complicated IVs might be?

jakobbotsch replied:
I haven't checked, but I have some (hacky) code lying around for it that I was planning to look at the impact of later.

There are some other notable missing things in the SCEV IR, like a subtraction operator, so loops that subtract by an invariant local are not representable (subtractions with a constant are usually represented as `GT_ADD` so those are ok).

AndyAyersMS commented on src/coreclr/jit/inductionvariableopts.cpp:
```diff
@@ -0,0 +1,1533 @@
+    }
+    else
+    {
+        return nullptr;
```
Maybe add a comment here noting that this null return means we do not have the simple IV shape (1 direct update of the def from the phi)?

AndyAyersMS commented on src/coreclr/jit/inductionvariableopts.cpp:
```diff
@@ -0,0 +1,1533 @@
+    }
+
+    return nullptr;
+}
```
Hoisting has a more refined notion of loop invariance that can handle more tree shapes; any thought of using something like that here?

jakobbotsch replied:
Do you mean by switching to use VN? I suppose that would be possible (although somewhat annoying since the VN wouldn't give me things like "reaching def from outside the loop" that I'm making use of here). I think it would mainly get us the ability to reason about loop invariant loads (so you could have e.g. `<L00, 0, IND(x)>`, provided we also started representing loads in the SCEV IR).

AndyAyersMS reviewed (commented):
Did a first review pass and it looks good overall. Will look again later.

I wonder if you might save some TP if you kept track of the uses during the locate phase and used that to drive replacement.

More generally if we want to recognize secondary IVs it seems like walking forward from the primary IV appearances might be a good plan. Perhaps when we get to strength reduction?

jakobbotsch committed:
Optimize throughput

jakobbotsch committed:
Clean up, address some feedback

jakobbotsch committed:
More feedback

AndyAyersMS reviewed (approved):
LGTM.

Will be interesting to see how this evolves when you start to look for secondary IVs.

BruceForstall commented:
> cc @dotnet/jit-contrib PTAL @BruceForstall @AndyAyersMS

I've started looking at this.

BruceForstall reviewed (approved):
Some comments:

1. It would be valuable for this code to have more design documentation. E.g., a Markdown document (in the JIT "doc" tree) on SCEV, on the IV widening algorithm design, etc. This could also take the form of more extensive design-oriented comments.
2. Would it make sense to extract the SCEV code to separate scev.h/cpp files? That is, separate the analysis and optimization code.
3. Add cScev/dScev debug dumpers?
4. Is there any useful dumper to be added to clrjt.natvis?
5. To match LLVM conventions, should the SCEV display the addition operator (even though that is currently the only one), e.g., `<loop, start, +, step>` or `<start, +, step>(loop)`?

jakobbotsch committed:
Split out into multiple files

jakobbotsch committed:
Address some feedback

jakobbotsch committed:
Add some docs

jakobbotsch commented:
> Some comments:
> 
> 1. It would be valuable for this code to have more design documentation. E.g., a Markdown document (in the JIT "doc" tree) on SCEV, on the IV widening algorithm design, etc. This could also take the form of more extensive design-oriented comments.

Makes sense, but personally I prefer to have it in the relevant files, so I added some documentation there.

> 2. Would it make sense to extract the SCEV code to separate scev.h/cpp files? That is, separate the analysis and optimization code.

Yeah, seems good to me. I split it into scev.h, scev.cpp and inductionvariableopts.cpp.

> 3. Add cScev/dScev debug dumpers?

Done.

> 4. Is there any useful dumper to be added to clrjt.natvis?

Added some basic display of the SCEV oper/type.

> 5. To match LLVM conventions, should the SCEV display the addition operator (even though that is currently the only one), e.g., `<loop, start, +, step>` or `<start, +, step>(loop)`?

I actually matched the notation from the paper (I think LLVM uses curly braces as well, and has a different representation of chains of recurrences). If we end up with more complex recurrences then we can consider changing the notation.

jakobbotsch commented on src/coreclr/jit/inductionvariableopts.cpp:
```diff
@@ -0,0 +1,1519 @@
+// Parameters:
+//   scev - The scev node.
+//
+void ScalarEvolutionContext::DumpScev(Scev* scev)
```
The main thing was that I wanted to display the specific loop for `ScevAddRec`, and `ScevAddRec` itself doesn't have the loop field. However, I just added that loop field as a debug-only field, and made the dumping a member method on `Scev`. So now we have `cScev` and `dScev`.

jakobbotsch committed:
More remarks

jakobbotsch commented on src/coreclr/jit/inductionvariableopts.cpp:
Renamed it to `optBestEffortReplaceNarrowIVUses`, and added a remarks:
```cpp
// Remarks:
//   This function is best effort; it might not find all uses of the provided
//   SSA num, particularly because it does not follow into joins. Note that we
//   only use this to replace uses of the narrow IV outside the loop; inside
//   the loop we do ensure that all uses/defs are replaced.
//   Keeping it best-effort outside the loop is ok; there is no correctness
//   issue since we do not invalidate the value of the old narrow IV in any
//   way, but it may mean we end up leaving the narrow IV live concurrently
//   with the new widened IV, increasing register pressure.
```

jakobbotsch committed:
Expand ryujit markdown docs with new phase

jakobbotsch commented:
SPMI 0 sized diffs look like #98996. All CI jobs were green last time it ran, and since then I only reorganized some things without any diffs, so the failures/timeouts look unrelated.
